# Funktionen in Python – Lernheft

## Von den Grundlagen bis Lambda, `*args`, `**kwargs`, Callback, Rekursion und Generatoren

Eine Funktion ist ein benannter, wiederverwendbarer Codeblock. Sie wird erst beim Aufruf ausgeführt.

Funktionen verbessern vor allem:

- Struktur und Übersicht,
- Wiederverwendbarkeit,
- Wartbarkeit,
- Testbarkeit,
- Anpassbarkeit,
- Vermeidung von doppeltem Code.

> **Wichtig:** Funktionen steigern nicht automatisch die Performance. Ihr Hauptvorteil ist die saubere Organisation von Code.

# 1. Funktion definieren und aufrufen

Allgemeine Form:

```python
def funktionsname():
    anweisungen
```

In [1]:
def begruessen():
    print("Hallo Holger!")


begruessen()

Hallo Holger!


Ohne den Aufruf `begruessen()` wird der Funktionskörper nicht ausgeführt.

# 2. Parameter und Argumente

Ein **Parameter** ist ein Platzhalter in der Definition.  
Ein **Argument** ist der konkrete Wert beim Aufruf.

In [2]:
def begruessen_person(name):
    print(f"Hallo {name}!")


begruessen_person("Holger")
begruessen_person("Rebecca")

Hallo Holger!
Hallo Rebecca!


Hier ist `name` der Parameter und `"Holger"` ein Argument.

# 3. Mehrere Parameter

In [3]:
def beschreibe_person(name, alter, stadt):
    print(f"{name} ist {alter} Jahre alt und lebt in {stadt}.")


beschreibe_person("Holger", 40, "Mönchengladbach")

Holger ist 40 Jahre alt und lebt in Mönchengladbach.


Positionsargumente werden von links nach rechts zugeordnet.

# 4. `print()` und `return`

`print()` zeigt nur etwas an.  
`return` gibt einen Wert an den aufrufenden Programmteil zurück.

In [4]:
def addieren_mit_print(a, b):
    print(a + b)


ergebnis = addieren_mit_print(4, 6)
print("Gespeicherter Rückgabewert:", ergebnis)

10
Gespeicherter Rückgabewert: None


In [5]:
def addieren(a, b):
    return a + b


ergebnis = addieren(4, 6)
print("Ergebnis:", ergebnis)
print("Doppeltes Ergebnis:", ergebnis * 2)

Ergebnis: 10
Doppeltes Ergebnis: 20


> **Merksatz:** `print()` ist für die Anzeige. `return` ist für die Weiterverarbeitung.

# 5. Mehrere `return`-Stellen

Sobald `return` ausgeführt wird, endet die Funktion.

In [6]:
def pruefe_alter(alter):
    if alter < 0:
        return "Ungültiges Alter"
    if alter < 18:
        return "Minderjährig"
    return "Volljährig"


for wert in (-2, 16, 25):
    print(wert, "->", pruefe_alter(wert))

-2 -> Ungültiges Alter
16 -> Minderjährig
25 -> Volljährig


# 6. Standardparameter

In [7]:
def begruessen_standard(name="Gast"):
    print(f"Hallo {name}!")


begruessen_standard()
begruessen_standard("Holger")

Hallo Gast!
Hallo Holger!


Pflichtparameter müssen vor Standardparametern stehen.

# 7. Schlüsselwortargumente

In [8]:
def beschreibe_person_kw(name, alter, stadt):
    print(f"{name} ist {alter} Jahre alt und lebt in {stadt}.")


beschreibe_person_kw(
    alter=40,
    stadt="Mönchengladbach",
    name="Holger"
)

Holger ist 40 Jahre alt und lebt in Mönchengladbach.


Bei Schlüsselwortargumenten spielt die Reihenfolge keine Rolle.

# 8. Mehrere Rückgabewerte

In [9]:
def rechnen(a, b):
    return a + b, a - b, a * b


werte = rechnen(10, 4)
print(werte)
print(type(werte))

summe, differenz, produkt = werte
print(summe, differenz, produkt)

(14, 6, 40)
<class 'tuple'>
14 6 40


Technisch wird ein Tupel zurückgegeben.

# 9. Lokale und globale Variablen

In [10]:
def lokale_demo():
    lokale_variable = "Nur innerhalb der Funktion sichtbar"
    print(lokale_variable)


lokale_demo()

Nur innerhalb der Funktion sichtbar


In [11]:
kursname = "Python-Grundlagen"


def zeige_kurs():
    print(kursname)


zeige_kurs()

Python-Grundlagen


Globale Variablen können gelesen werden. Änderungen mit `global` sind möglich, aber meist schlechter nachvollziehbar als Parameter und Rückgabewerte.

In [12]:
zaehler = 0


def erhoehe_zaehler():
    global zaehler
    zaehler += 1


erhoehe_zaehler()
erhoehe_zaehler()
print(zaehler)

2


# 10. Docstrings

In [13]:
def berechne_flaeche(laenge, breite):
    """Berechnet die Fläche eines Rechtecks."""
    return laenge * breite


print(berechne_flaeche(5, 3))
print(berechne_flaeche.__doc__)

15
Berechnet die Fläche eines Rechtecks.


Ein Docstring erklärt Zweck, Parameter, Rückgabewert und gegebenenfalls mögliche Fehler.

# 11. Typannotationen

In [14]:
def multiplizieren(a: int, b: int) -> int:
    return a * b


print(multiplizieren(4, 5))

20


Typannotationen sind Hinweise. Python erzwingt sie zur Laufzeit nicht automatisch.

# 12. Lambda-Funktionen

Eine Lambda-Funktion ist eine kurze anonyme Funktion.

```python
lambda parameter: ausdruck
```

In [15]:
quadrat = lambda x: x ** 2
print(quadrat(5))

25


In [16]:
def quadrat_def(x):
    return x ** 2


print(quadrat_def(5))

25


Lambda eignet sich für kurze, einfache Ausdrücke. Bei komplexerer Logik ist `def` besser lesbar.

# 13. Lambda mit `sorted()`

In [17]:
personen = [
    {"name": "Anna", "alter": 25},
    {"name": "Ben", "alter": 19},
    {"name": "Chris", "alter": 32},
]

sortiert = sorted(personen, key=lambda person: person["alter"])
print(sortiert)

[{'name': 'Ben', 'alter': 19}, {'name': 'Anna', 'alter': 25}, {'name': 'Chris', 'alter': 32}]


Die Lambda-Funktion liefert für jede Person den Sortierschlüssel.

# 14. Lambda mit `map()` und `filter()`

In [18]:
zahlen = [1, 2, 3, 4, 5]

verdoppelt = list(map(lambda zahl: zahl * 2, zahlen))
gerade = list(filter(lambda zahl: zahl % 2 == 0, zahlen))

print("Verdoppelt:", verdoppelt)
print("Gerade:", gerade)

Verdoppelt: [2, 4, 6, 8, 10]
Gerade: [2, 4]


In Python sind List Comprehensions häufig leichter lesbar:

In [19]:
verdoppelt_lc = [zahl * 2 for zahl in zahlen]
gerade_lc = [zahl for zahl in zahlen if zahl % 2 == 0]

print(verdoppelt_lc)
print(gerade_lc)

[2, 4, 6, 8, 10]
[2, 4]


# 15. `*args`

Mit `*args` kann eine Funktion beliebig viele Positionsargumente erhalten. `args` ist ein Tupel.

In [20]:
def addiere_beliebig(*zahlen):
    print("Tupel:", zahlen)
    return sum(zahlen)


print(addiere_beliebig(1, 2, 3))
print(addiere_beliebig(10, 20, 30, 40))

Tupel: (1, 2, 3)
6
Tupel: (10, 20, 30, 40)
100


# 16. `**kwargs`

Mit `**kwargs` kann eine Funktion beliebig viele Schlüsselwortargumente erhalten. `kwargs` ist ein Dictionary.

In [21]:
def zeige_person(**daten):
    print("Dictionary:", daten)
    for schluessel, wert in daten.items():
        print(f"{schluessel}: {wert}")


zeige_person(
    name="Holger",
    alter=40,
    stadt="Mönchengladbach"
)

Dictionary: {'name': 'Holger', 'alter': 40, 'stadt': 'Mönchengladbach'}
name: Holger
alter: 40
stadt: Mönchengladbach


# 17. `*args` und `**kwargs` gemeinsam

In [22]:
def zeige_werte(titel, *args, **kwargs):
    print(titel)
    print("Positionsargumente:", args)
    print("Schlüsselwortargumente:", kwargs)


zeige_werte("Beispiel", 1, 2, 3, name="Holger", aktiv=True)

Beispiel
Positionsargumente: (1, 2, 3)
Schlüsselwortargumente: {'name': 'Holger', 'aktiv': True}


# 18. Funktionen sind Werte

In [23]:
def verdoppeln(zahl):
    return zahl * 2


funktion = verdoppeln
print(funktion(5))

10


`funktion = verdoppeln` speichert die Funktion selbst.  
`funktion = verdoppeln(5)` führt sie sofort aus und speichert das Ergebnis.

# 19. Callback-Funktionen

Ein Callback ist eine Funktion, die an eine andere Funktion übergeben wird.

In [24]:
def ausfuehren(funktion, wert):
    return funktion(wert)


def verdreifachen(zahl):
    return zahl * 3


print(ausfuehren(verdreifachen, 5))
print(ausfuehren(lambda zahl: zahl + 100, 25))

15
125


# 20. Rekursion

Eine rekursive Funktion ruft sich selbst auf. Jede Rekursion braucht eine Abbruchbedingung.

In [25]:
def fakultaet(n):
    if n <= 1:
        return 1
    return n * fakultaet(n - 1)


print(fakultaet(5))

120


`5! = 5 × 4 × 3 × 2 × 1 = 120`

Ohne Abbruchbedingung entsteht irgendwann ein `RecursionError`.

# 21. Generator-Funktionen

Generator-Funktionen verwenden `yield`. Sie liefern Werte schrittweise.

In [26]:
def zahlen_bis(limit):
    for zahl in range(limit + 1):
        yield zahl


generator = zahlen_bis(5)
print(generator)
print(list(generator))

<generator object zahlen_bis at 0x7ed93dade1f0>
[0, 1, 2, 3, 4, 5]


Generatoren sind bei großen Datenmengen, Dateien und Datenströmen sinnvoll.

# 22. Funktion und Methode unterscheiden

Eine Funktion steht eigenständig:

```python
funktion()
```

Eine Methode gehört zu einer Klasse oder einem Objekt:

```python
objekt.methode()
```

In [27]:
def grossschreiben(text):
    return text.upper()


name = "holger"

print(grossschreiben(name))
print(name.upper())

HOLGER
HOLGER


In [28]:
liste = [3, 1, 2]
liste.append(4)
liste.sort()

dictionary = {"name": "Holger"}

print(liste)
print(dictionary.get("name"))

[1, 2, 3, 4]
Holger


- `grossschreiben()` ist eine Funktion.
- `.upper()` ist eine String-Methode.
- `.append()` und `.sort()` sind Listenmethoden.
- `.get()` ist eine Dictionary-Methode.

> **Merksatz:** Jede Methode ist eine Funktion, aber nicht jede Funktion ist eine Methode.

# 23. Typische Fehler

- Funktion definiert, aber nicht aufgerufen
- zu wenige oder zu viele Argumente
- `print()` statt `return`
- Pflichtparameter hinter Standardparametern
- globale Variablen unnötig verändern
- Lambda für zu komplexe Logik verwenden
- Rekursion ohne Abbruchbedingung

# 24. Übung 1 – Begrüßungsfunktion

Erstelle eine Funktion, die Name und Stadt erhält und einen Begrüßungstext zurückgibt.

In [29]:
def begruessung(name, stadt):
    return f"Hallo {name} aus {stadt}!"


print(begruessung("Holger", "Mönchengladbach"))

Hallo Holger aus Mönchengladbach!


# 25. Übung 2 – Bruttopreis

In [30]:
def berechne_brutto(nettopreis, steuersatz_prozent=19):
    faktor = 1 + steuersatz_prozent / 100
    return round(nettopreis * faktor, 2)


print(berechne_brutto(100))
print(berechne_brutto(100, 7))

119.0
107.0


# 26. Übung 3 – Gerade oder ungerade

In [31]:
def gerade_oder_ungerade(zahl):
    if zahl % 2 == 0:
        return "gerade"
    return "ungerade"


print(gerade_oder_ungerade(8))
print(gerade_oder_ungerade(7))

gerade
ungerade


# 27. Übung 4 – Größter Wert mit `*args`

In [32]:
def groesster_wert(*zahlen):
    if not zahlen:
        return None
    return max(zahlen)


print(groesster_wert(3, 7, 2, 10, 5))
print(groesster_wert())

10
None


# 28. Übung 5 – Person mit `**kwargs`

In [33]:
def personenbeschreibung(**person):
    name = person.get("name", "Unbekannt")
    alter = person.get("alter", "unbekannt")
    stadt = person.get("stadt", "unbekannt")
    return f"{name} ist {alter} Jahre alt und lebt in {stadt}."


print(
    personenbeschreibung(
        name="Holger",
        alter=40,
        stadt="Mönchengladbach"
    )
)

Holger ist 40 Jahre alt und lebt in Mönchengladbach.


# Zusammenfassung

Behandelt wurden:

- `def` und Funktionsaufruf
- Parameter und Argumente
- `print()` und `return`
- Standardparameter
- Schlüsselwortargumente
- mehrere Rückgabewerte
- lokale und globale Variablen
- Docstrings und Typannotationen
- Lambda
- `map()` und `filter()`
- `*args` und `**kwargs`
- Funktionen als Werte
- Callback
- Rekursion
- Generatoren mit `yield`
- Funktion und Methode
- typische Fehler

## Kurz-Check

1. Was unterscheidet Parameter und Argument?
2. Was unterscheidet `print()` und `return`?
3. Wann wird eine Funktion ausgeführt?
4. Was ist ein Standardparameter?
5. Was wird bei mehreren Rückgabewerten technisch zurückgegeben?
6. Was ist eine lokale Variable?
7. Was ist eine Lambda-Funktion?
8. Was unterscheidet `*args` und `**kwargs`?
9. Was ist ein Callback?
10. Was braucht jede Rekursion?
11. Was unterscheidet `yield` und `return`?
12. Was unterscheidet Funktion und Methode?